### PR TITLE
Simplify constant Expr

### PR DESCRIPTION
**Pull request by kangkaisen**
Simplify constant Expr could improve Partition Pruning.   The examples for constant Expr Simplifing:
```
1 + 1 + 1 --> 3
date_add('2018-08-08', 1) --> 2018-08-09
year('2018-07-24')*12 + month('2018-07-24') -> 24223
```

This implementation VS Impala implementation:

The prons:

**No RPC, the performance is better.**

The cons:

**Need to implement these functions in FE again, which need a lot of work, but we could implement these functions one by one and according to the users need like String functions.**

### CODE REVIEW

**Comment by kangkaisen** [comment]
Update this commit.

File: fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+            if (!Arrays.equals(argTypes1, argTypes2)) {
+                continue;
+            }
+
+            return invoker;
+        }
+        return null;
+    }
+
+    private void registerFunctions() {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java]: This should be synchronized? 
Now, there may concurrent calling

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java (hunk 1 above)
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java]: Yes.  Fixed.

File: fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+            if (!Arrays.equals(argTypes1, argTypes2)) {
+                continue;
+            }
+
+            return invoker;
+        }
+        return null;
+    }
+
+    private synchronized void registerFunctions() {
```
Comment (imay) [inline fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java]: first you need to check if the map is empty() too. Other one function will be added multiple times

**Review by imay — CHANGES_REQUESTED** [review]

File: fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java (hunk 2 above)
Comment (kangkaisen) [reply fe/src/main/java/org/apache/doris/rewrite/FoldConstantsRule.java]: Yes. which like singleton-pattern. I have updated. 
But in fact, even if one function added multiple times, the `getFunction` would still get a right Function.

**Review by imay — APPROVED** [review]

**Pull request merged**